things/api/http/endpoints_test.go

```diff
@@ -1947,12 +1940,826 @@ func TestListMembers(t *testing.T) {
+
+	cases := []struct {
+		desc        string
+		token       string
+		channelID   string
+		thingID     string
+		contentType string
+		status      int
+		err         error
+	}{
+		{
+			desc:        "disconnect thing from channel successfully",
+			token:       validToken,
+			channelID:   validID,
+			thingID:     validID,
+			contentType: contentType,
+			status:      http.StatusNoContent,
+			err:         nil,
+		},
+		{
+			desc:        "disconnect thing from channel with invalid token",
+			token:       inValidToken,
+			channelID:   validID,
+			thingID:     validID,
+			contentType: contentType,
+			status:      http.StatusUnauthorized,
+			err:         svcerr.ErrAuthentication,
+		},
+		{
+			desc:        "disconnect thing from channel with empty channel id",
+			token:       validToken,
+			channelID:   "",
+			thingID:     validID,
+			contentType: contentType,
+			status:      http.StatusUnprocessableEntity,
+			err:         apiutil.ErrValidation,
+		},
+		{
+			desc:        "disconnect thing from channel with empty thing id",
+			token:       validToken,
+			channelID:   validID,
+			thingID:     "",
+			contentType: contentType,
+			status:      http.StatusUnprocessableEntity,
+			err:         apiutil.ErrValidation,
+		},
+	}
+	for _, tc := range cases {
+		req := testRequest{
+			client:      ts.Client(),
+			method:      http.MethodPost,
+			url:         fmt.Sprintf("%s/channels/%s/things/%s/disconnect", ts.URL, tc.channelID, tc.thingID),
+			token:       tc.token,
+			contentType: tc.contentType,
+		}
+
+		svcCall := gsvc.On("Unassign", mock.Anything, tc.token, tc.channelID, "group", "things", []string{tc.thingID}).Return(tc.err)
+		res, err := req.make()
+		assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+		assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
+		svcCall.Unset()
+	}
+}
+
+func TestConnect(t *testing.T) {
+	ts, _, gsvc := newThingsServer()
+	defer ts.Close()
+
+	cases := []struct {
+		desc        string
+		token       string
+		reqBody     interface{}
+		contentType string
+		status      int
+		err         error
+	}{
+		{
+			desc:  "connect thing to channel successfully",
+			token: validToken,
+			reqBody: groupReqBody{
+				ChannelID: validID,
+				ThingID:   validID,
+			},
+			contentType: contentType,
+			status:      http.StatusCreated,
+			err:         nil,
+		},
+		{
+			desc:  "connect thing to channel with invalid token",
+			token: inValidToken,
+			reqBody: groupReqBody{
+				ChannelID: validID,
+				ThingID:   validID,
+			},
+			contentType: contentType,
+			status:      http.StatusUnauthorized,
+			err:         svcerr.ErrAuthentication,
+		},
+		{
+			desc:  "connect thing to channel with empty channel id",
+			token: validToken,
+			reqBody: groupReqBody{
+				ChannelID: "",
+				ThingID:   validID,
+			},
+			contentType: contentType,
+			status:      http.StatusUnprocessableEntity,
```

Comment:
This should be code 400 change validation to return missing id. Same as other cases